chore: remove graphql-utils and allow for the use of swc (#2176)
vtex/faststore: apps/site/pages/docs/migration-guides/Migrating to 2.2.65 - graphql-utils depreciation.mdx

# Migrating to 2.2.65 - graphql-utils depreciation

> From 2.2.60 on, the `@faststore/graphql-utils` is officially deprecated. It was previously used internally by `@faststore/core` to handle GraphQL optimizations and GraphQL definitions.

Reviewer: 🎗️ remember to update it to the correct version.

> From 2.2.60 on, the `@faststore/graphql-utils` is officially deprecated. It was previously used internally by `@faststore/core` to handle GraphQL optimizations and GraphQL definitions.

Reviewer 2: ```suggestion
In this guide, learn how to migrate your store version to v2.2.65 to leverage the latest improvements in the API extension.

Version 2.2.65 and above, introduces the following enhancements to API extension users: 

- Deprecation of the `@faststore/graphql-utils` package in favor of the [`client-preset`](https://the-guild.dev/graphql/codegen/plugins/presets/preset-client) plugin.

- Refinement `gql` helper usage for a smoother and more efficient GraphQL query handling.

- Enhancement security measures to a more secure method for invoking queries and mutations to safeguard your store's data.

- Optimization call processes for queries or mutations defined within `@faststore/core`.

import { Callout } from 'nextra/components'
 
<Callout type="info" emoji="ℹ️">
  For more details about these changes, also refer to the release note [TBD](/tbd)
</Callout>
```

Reviewer 2: The guide is important to have, but it appears that a release note is also necessary. This note will officially inform CX about the migration and enhance visibility for the guide. I'm currently drafting the release note [here](https://docs.google.com/document/d/1HZINqZ0f-iDQb3CnuyPki2A74RmJ0Rbqr4Ud5S1Fi6Q/edit?usp=sharing), so the team can also review it.

Author: I agree! Thanks for that.

> The `@faststore/graphql-utils` has been replaced by open-source solutions maintained by the community that are now re-exported from `@faststore/core`. There are minor breaking changes on how developers should write GraphQL definitions and invoke queries and mutations, which were introduced in version 2.2.60.

Reviewer: 🎗️  also here regarding the version

> 

Reviewer 2: ```suggestion
## Before you begin
Make sure your store version is updated to v2.2.65 or above. If it’s not updated follow the instructions below:

1. Open your store code and navigate to the `package.json` file.

2. In `dependencies` > `@faststore/core`, change the version to the following:

```json
"@faststore/core": "^2.2.65",
```

3. Open the terminal and run `yarn` to update the version.

```

> ## The `gql` helper

Reviewer 2: ```suggestion
## Updating the `gql` helper usage
```

> Previously, it was possible to import the `gql` helper function directly from `@faststore/graphql-utils` - although it was not recommended. This will now throw an error. See the example below:

Reviewer 2: Why was not recommended?

> Previously, it was possible to import the `gql` helper function directly from `@faststore/graphql-utils` - although it was not recommended. This will now throw an error. See the example below:

Author: Because users should import everything from @faststore/core, that's the centralized package that should include everything you need to build a store. Importing the same code from multiple sources may cause versioning errors when handling dependencies.

```tsx
// src/fragments/ClientProduct.tsx
import { gql } from '@faststore/graphql-utils'

export const fragment = gql`
  fragment ClientProduct on Query {
    product(locator: $locator) {
      id: productID
      breadcrumbList {
        itemListElement {
          item
          name
          position
        }
      }
    }
  }
`
```

> The code above should be re-written to accomodate the changes in the `gql` helper, which should be imported from `@faststore/core/api` and be invoked as a function - with the argument between parentesis. This also applies to query and mutation definitions inside the components. See the re-written code below:

Reviewer 2: ```suggestion
Now with the  v2.2.65, you should import the `gql` helper from `@faststore/core/api`and be called as a function - with the argument between parenthesis. This also applies to query and mutation definitions inside the components. For example:
```

> ```tsx

Reviewer 2: ```suggestion
```tsx {1}
```

// src/fragments/ClientProduct.tsx
import { gql } from '@faststore/core/api'

export const fragment = gql(`
  fragment ClientProduct on Query {
    product(locator: $locator) {
      id: productID
      breadcrumbList {
        itemListElement {
          item
          name
          position
        }
      }
    }
  }
`)
```

> ## The `useQuery` hook

Reviewer 2: ```suggestion
## Using `useQuery` hook to call queries and mutations
```

> Another change was made on how queries and mutations were invoked. Previously, it was possible to invoke queries and mutations by using their names - such as `MyCustomQuery` in the example below - just by providing it to the `useQuery` hook. This is no longer possible, as queries and mutations are now invoked using more secure hashes which are randomly generated. Here's an example:

Reviewer 2: ```suggestion
Previously, it was possible to invoke queries and mutations by using their names - such as `MyCustomQuery` by providing it to the `useQuery` hook. For example:
```

```tsx
import { useQuery_unstable as useQuery } from '@faststore/core/experimental'

const query = gql(`
  query MyCustomQuery {
    customQuery() {
      data
    }
  }
`)

// The useQuery call will now throw an error
function CustomComponent() {
    // ...
    const result = useQuery(`MyCustomQuery`, variables)
    // ...
}
```

> To fix it, developers should pass the query or mutation object - result of the `gql` call -  to `useQuery` directly.

Reviewer 2: ```suggestion
With v2.2.65, queries and mutations are now only invoked using more secure hashes, which are randomly generated and do to that you must pass the query or mutation object - result of the `gql` call - to `useQuery` directly. For example:
```

```tsx
import { gql } from '@faststore/core/api'
import { useQuery_unstable as useQuery } from '@faststore/core/experimental'

const query = gql(`
  query MyCustomQuery {
    customQuery() {
      data
    }
  }
`)

// useQuery apropriately calls MyCustomQuery
function CustomComponent() {
    // ...
    const result = useQuery(query, variables)
    // ...
}
```

> ### Calling a query or mutation defined by `@faststore/core`

Reviewer 2: ```suggestion
### Calling queries or mutations defined by `@faststore/core`
```

> Sometimes a custom component needs to call a query or mutation defined by `@faststore/core`, such as `ClientManyProductsQuery`. In those cases, developers should replace the `useQuery` hook call by a call to the specific hook for that query. The availability for such hooks is limited, and we're working to provide support for them. See the example below.
> 

Reviewer 2: ```suggestion
A custom component can call a query or mutation defined by `@faststore/core`, such as `ClientManyProductsQuery`. In these cases, you replace the `useQuery` hook call with a call to the specific hook for that query. 

<Callout type="warning" emoji="⚠️">
  The availability of such hooks is limited.
</Callout>
```

Reviewer 2: I didn't understand about the availability limitation for these hooks. 🤔
What kind of limitations? Just want to know if we need to be more specific and clear here.

Author: We just haven't created them yet.

> # Migrating your store to v2.2.65

Reviewer 2: ```suggestion
# Improvements to API extensions 
```

Reviewer 2: @icazevedo, @hellofanny and I were talking about the word "Migrating" in this context, and maybe it can cause some misunderstanding. Maybe people will think we are talking about a new FastStore version and not exactly the faststore/core package. What do you think about this new title suggestion?

> In this guide, learn how to migrate your store version to v2.2.65 to leverage the latest improvements in the API extension.

Reviewer 2: ```suggestion
In this guide, learn how to migrate your store version to v3.0.0 to leverage the latest improvements in the API extension.
```

> Version 2.2.65 and above, introduces the following enhancements to API extension users: 

Reviewer 2: ```suggestion
Version 3.0.0 and above, introduces the following enhancements to API extension users: 
```

- Deprecation of the `@faststore/graphql-utils` package in favor of the [`client-preset`](https://the-guild.dev/graphql/codegen/plugins/presets/preset-client) plugin.

- Refinement `gql` helper usage for a smoother and more efficient GraphQL query handling.

- Enhancement security measures to a more secure method for invoking queries and mutations to safeguard your store's data.

- Optimization call processes for queries or mutations defined within `@faststore/core`.

import { Callout } from 'nextra/components'
 
<Callout type="info" emoji="ℹ️">

>   For more details about these changes, also refer to the release note [TBD](/tbd)

Reviewer 2: ```suggestion
  For more details about these changes, also refer to the [GitHub releases](/tbd) related to this version.
```

</Callout>

> The `@faststore/graphql-utils` has been replaced by open-source solutions maintained by the community that are now re-exported from `@faststore/core`. There are minor breaking changes on how developers should write GraphQL definitions and invoke queries and mutations, which were introduced in version 2.2.60.

Reviewer 2: ```suggestion
The `@faststore/graphql-utils` has been replaced by open-source solutions maintained by the community that are now re-exported from `@faststore/core`. There are minor breaking changes on how developers should write GraphQL definitions and invoke queries and mutations, which were introduced in version 3.0.0.
```

## Before you begin

> Make sure your store version is updated to v2.2.65 or above. If it’s not updated follow the instructions below:

Reviewer 2: ```suggestion
Make sure your store version is updated to v3.0.0 or above. If it’s not updated follow the instructions below:
```

1. Open your store code and navigate to the `package.json` file.

2. In `dependencies` > `@faststore/core`, change the version to the following:

```json

> "@faststore/core": "^2.2.65",

Reviewer 2: ```suggestion
"@faststore/core": "^3.0.0",
```

```

> 3. Open the terminal and run `yarn` to update the version.

Reviewer 2: ```suggestion
3. Open the terminal and run `yarn` to update the version.

```

## Updating the `gql` helper usage

The `gql` helper serves as a function to define GraphQL operations such as queries, mutations, or fragments within the store code. Before, the function was imported directly from the `@faststore/graphql-utils` which was not recommended. See the example below:

> ```tsx
> // src/fragments/ClientProduct.tsx

Reviewer 3: ```suggestion
```tsx filename="src/fragments/ClientProduct.tsx"
```

import { gql } from '@faststore/graphql-utils'

export const fragment = gql`
  fragment ClientProduct on Query {
    product(locator: $locator) {
      id: productID
      breadcrumbList {
        itemListElement {
          item
          name
          position
        }
      }
    }
  }
`
```

> Now with the  v2.2.65, you should import the `gql` helper from `@faststore/core/api`and be called as a function - with the argument between parenthesis. This also applies to query and mutation definitions inside the components. For example:

Reviewer 2: ```suggestion
Now with the  v3.0.0, you should import the `gql` helper from `@faststore/core/api`and be called as a function - with the argument between parenthesis. This also applies to query and mutation definitions inside the components. For example:
```

> ```tsx {1}
> // src/fragments/ClientProduct.tsx

Reviewer 3: ```suggestion
```tsx filename="src/fragments/ClientProduct.tsx" {1}
```

import { gql } from '@faststore/core/api'

export const fragment = gql(`
  fragment ClientProduct on Query {
    product(locator: $locator) {
      id: productID
      breadcrumbList {
        itemListElement {
          item
          name
          position
        }
      }
    }
  }
`)
```

## Using `useQuery` hook to call queries and mutations

Previously, it was possible to invoke queries and mutations by using their names - such as `MyCustomQuery` by providing it to the `useQuery` hook. For example:

```tsx
import { useQuery_unstable as useQuery } from '@faststore/core/experimental'

const query = gql(`
  query MyCustomQuery {
    customQuery() {
      data
    }
  }
`)

// The useQuery call will now throw an error
function CustomComponent() {
    // ...
    const result = useQuery(`MyCustomQuery`, variables)
    // ...
}
```

> With v2.2.65, queries and mutations are now only invoked using more secure hashes, which are randomly generated and do to that you must pass the query or mutation object - result of the `gql` call - to `useQuery` directly. For example:

Reviewer 2: ```suggestion
With v3.0.0, queries and mutations are now only invoked using more secure hashes, which are randomly generated and do to that you must pass the query or mutation object - result of the `gql` call - to `useQuery` directly. For example:
```
